locallib.php

```diff
@@ -443,21 +443,51 @@ function local_boostnavigation_cohort_is_member($userid, $setting) {
     // Initialize variable for memberships.
     static $allmemberships = null;
 
-    // First: If the memberships haven't been fetched yet initially, fetch all of the user's cohort memberships
-    // only once and remember for next calls of this function.
-    if ($allmemberships == null) {
-        // Prepare SQL statement.
-        $sql = 'SELECT {cohort}.id, {cohort}.idnumber FROM {cohort_members}
-            JOIN {cohort}
-                ON {cohort}.id = {cohort_members}.cohortid
-            WHERE {cohort_members}.userid = ?';
-        $params = array($userid);
-
-        // Run DB query.
-        $ret = $DB->get_records_sql_menu($sql, $params);
-
-        // Remember memberships.
-        $allmemberships = $ret;
+    try {
+        $cohortversioncache = cache::make('local_boostnavigation', 'cohortversion');
+        $usercohortversioncache = cache::make('local_boostnavigation', 'usercohortversion');
+        $usercohortscache = cache::make('local_boostnavigation', 'usercohorts');
+    } catch (Exception $e) {
+        throw $e;
+    }
+
+    // Set cohort version cache with current timestamp if the cohort version is not set
+    // from the cohort event observers.
+    if (!$cohortversioncache->get('cohortversion')) {
+        $cohortversioncache->set('cohortversion', time());
+    }
+
+    // Check if the user cohort version and cohort version cache are not same, then we need to read cohorts from db
+    if (!$usercohortscache->get('usercohorts') ||
```

Comment:
I think this can be a lot simpler. The idea is fleshed out in this note:

https://wrms.catalyst.net.nz/wr.php?request_id=319065#note-9

You should only need 2 caches, 1 is the version number, and one is the cohort cache in the users session. 

First grab the global $cohortversion, lets say it's value is 123456789

Then you ask for 'that version' of the users cohort memberships something like:

$allmemberships = $usercohortscache->get('usercohorts' . $cohortversion);

This cache key / value will never change once set, it will either exist or not. Because it is immutable it can be localized in it's definition:

 'canuselocalstore' => true

https://docs.moodle.org/dev/Cache_API#The_definition

Reply:
Thank you Brendan !!

I really appreciate your taking the time to do this review. I've fixed these issues and pushed changes for you to review.

Thanks,
Anu